File: BandwidthScheduler/bandwidthscheduler.client/src/app/DataStructures/Heap.ts
export class Heap<T> {
  private readonly _elements: T[];
  private readonly _correctOrderCompare: (first: T, second: T) => boolean;

  constructor(compare: (first: T, second: T) => boolean) {
    this._elements = [];
    this._correctOrderCompare = compare;
  }

  private GetLeftChildIndex(elementIndex: number): number {
    return 2 * elementIndex + 1;
  }

  private GetRightChildIndex(elementIndex: number): number {
    return 2 * elementIndex + 2;
  }

  private GetParentIndex(elementIndex: number): number {
    return Math.floor((elementIndex - 1) / 2);
  }

  private HasLeftChild(elementIndex: number): boolean {
    return this.GetLeftChildIndex(elementIndex) < this._elements.length;
  }

  private HasRightChild(elementIndex: number): boolean {
    return this.GetRightChildIndex(elementIndex) < this._elements.length;
  }

  private IsRoot(elementIndex: number) {
    return elementIndex === 0;
  }

  private GetLeftChild(elementIndex: number): T {
    return this._elements[this.GetLeftChildIndex(elementIndex)];
  }

  private GetRightChild(elementIndex: number): T {
    return this._elements[this.GetRightChildIndex(elementIndex)];
  }

  private GetParent(elementIndex: number): T {
    return this._elements[this.GetParentIndex(elementIndex)];
  }

  private Swap(firstIndex: number, secondIndex: number): void {
    var temp = this._elements[firstIndex];
    this._elements[firstIndex] = this._elements[secondIndex];
    this._elements[secondIndex] = temp;
  }

  public Length(): number {
    return this._elements.length;
  }

  public Peek(): T {
    if (this._elements.length == 0) throw new RangeError();

    return this._elements[0];
  }

  public Pop(): T {
    if (this._elements.length == 0) throw new RangeError();

    var result = this._elements[0];
    this._elements[0] = this._elements[this._elements.length - 1];
    this._elements.splice(this._elements.length - 1, 1);

    this.ReCalculateDown(0);
    return result;
  }

  public Add(element: T): void {
    this._elements.push(element);

    this.ReCalculateUp(this._elements.length - 1);
  }

  private ReCalculateDown(index: number): number {
    while (this.HasLeftChild(index)) {
      let smallerIndex = this.GetLeftChildIndex(index);
      if (
        this.HasRightChild(index) &&
        this._correctOrderCompare(
          this.GetRightChild(index),
          this.GetLeftChild(index)
        )
      ) {
        smallerIndex = this.GetRightChildIndex(index);
      }

      if (
        !this._correctOrderCompare(
          this._elements[smallerIndex],
          this._elements[index]
        )
      ) {
        break;
      }

      this.Swap(smallerIndex, index);
      index = smallerIndex;
    }

    return index;
  }

  private ReCalculateUp(index: number): number {
    while (
      !this.IsRoot(index) &&
      this._correctOrderCompare(this._elements[index], this.GetParent(index))
    ) {
      const parentIndex = this.GetParentIndex(index);
      this.Swap(parentIndex, index);
      index = parentIndex;
    }

    return index;
  }
}

File: BandwidthScheduler/BandwidthScheduler.Server/Common/DataStructures/Heap.cs
namespace BandwidthScheduler.Server.Common.DataStructures
{
    public class Heap<T>
    {
        protected readonly List<T> _elements;
        protected readonly SequencingOrder _correctOrderCompare;

        public delegate bool SequencingOrder(T first, T second);
        public Heap(SequencingOrder comparison, int size = 0)
        {
            _correctOrderCompare = comparison;
            _elements = new List<T>(size);
        }

        private int GetLeftChildIndex(int elementIndex) => 2 * elementIndex + 1;
        private int GetRightChildIndex(int elementIndex) => 2 * elementIndex + 2;
        private int GetParentIndex(int elementIndex) => (elementIndex - 1) / 2;

        private bool HasLeftChild(int elementIndex) => GetLeftChildIndex(elementIndex) < _elements.Count;
        private bool HasRightChild(int elementIndex) => GetRightChildIndex(elementIndex) < _elements.Count;
        private bool IsRoot(int elementIndex) => elementIndex == 0;

        private T GetLeftChild(int elementIndex) => _elements[GetLeftChildIndex(elementIndex)];
        private T GetRightChild(int elementIndex) => _elements[GetRightChildIndex(elementIndex)];
        private T GetParent(int elementIndex) => _elements[GetParentIndex(elementIndex)];

        protected virtual void Swap(int firstIndex, int secondIndex)
        {
            var first = _elements[firstIndex];
            _elements[firstIndex] = _elements[secondIndex];
            _elements[secondIndex] = first;
        }

        public bool IsEmpty()
        {
            return _elements.Count == 0;
        }

        public int Count() => _elements.Count;

        public T Peek()
        {
            if (_elements.Count == 0)
                throw new IndexOutOfRangeException();

            return _elements[0];
        }

        public virtual T Pop()
        {
            if (_elements.Count == 0)
                throw new IndexOutOfRangeException();

            var result = _elements[0];
            _elements[0] = _elements[_elements.Count - 1];
            _elements.RemoveAt(_elements.Count - 1);

            ReCalculateDown(0);

            return result;
        }

        public virtual void Add(T element)
        {
            _elements.Add(element);

            ReCalculateUp(_elements.Count - 1);
        }

        protected int ReCalculateDown(int index)
        {
            while (HasLeftChild(index))
            {
                var smallerIndex = GetLeftChildIndex(index);
                if (HasRightChild(index) && _correctOrderCompare(GetRightChild(index), GetLeftChild(index)))
                {
                    smallerIndex = GetRightChildIndex(index);
                }

                if (!_correctOrderCompare(_elements[smallerIndex], _elements[index]))
                {
                    break;
                }

                Swap(smallerIndex, index);
                index = smallerIndex;
            }

            return index;
        }

        protected void ReCalculateUp(int index)
        {
            while (!IsRoot(index) && _correctOrderCompare(_elements[index], GetParent(index)))
            {
                var parentIndex = GetParentIndex(index);
                Swap(parentIndex, index);
                index = parentIndex;
            }
        }
    }
}
